File: mobile/ios/Enkodu/Enkodu/Transfer/TransferManager.swift
import Foundation
import CoreData

@MainActor
class TransferManager: ObservableObject {
    static let shared = TransferManager()

    private let api: EnkoduApi
    private let context: NSManagedObjectContext
    private let chunkSize: Int64 = 8 * 1024 * 1024 // 8 MiB
    private let maxRetries = 10
    private let baseDelayMs: Double = 500
    private let maxDelayMs: Double = 30000
    private let backoffMultiplier = 1.5

    @Published var activeTransfers: [TransferState] = []

    private init() {
        let serverURL = UserDefaults.standard.string(forKey: "serverURL") ?? "https://example.invalid"
        self.api = EnkoduApi(serverURL: serverURL)
        self.context = PersistenceController.shared.container.viewContext
    }

    // MARK: - Background Upload

    func uploadFileBackground(file: URL, onProgress: @escaping (Int64, Int64) -> Void = { _, _ in }, completion: @escaping (Result<UploadFinishResponse, Error>) -> Void) {
        let totalSize = (try? file.resourceValues(forKeys: [.fileSizeKey]).fileSize) ?? 0
        let serverURL = UserDefaults.standard.string(forKey: "serverURL") ?? "https://example.invalid"

        Task {
            do {
                let start = try await api.startResumableUpload(
                    filename: file.lastPathComponent,
                    filepath: file.path,
                    totalSize: Int64(totalSize)
                )
                let uploadId = start.uploadId

                // Save state for resume
                let state = TransferState(context: context)
                state.id = UUID()
                state.uploadId = uploadId
                state.filePath = file.path
                state.totalBytes = Int64(totalSize)
                state.status = TransferStatus.pending.rawValue
                state.transferType = TransferType.upload.rawValue
                state.createdAt = Date()
                state.updatedAt = Date()
                PersistenceController.shared.save()
                TelemetryManager.shared.trackUploadStart(jobId: uploadId)

                // Use background session for the actual upload
                let url = URL(string: "\(serverURL)/jobs/upload/resumable/\(uploadId)/chunk")!
                BackgroundSessionManager.shared.uploadFile(
                    url: url,
                    file: file,
                    uploadId: uploadId,
                    completion: { [weak self] in
                        // When all chunks are done, finish
                        Task {
                            do {
                                let finish = try await self?.api.finishResumableUpload(uploadId: uploadId)
                                if let finish = finish {
                                    TelemetryManager.shared.trackUploadSuccess(jobId: uploadId, durationMs: 0, bytesTransferred: Int64(totalSize))
                                    completion(.success(finish))
                                }
                            } catch {
                                TelemetryManager.shared.trackUploadFailure(error: "Background finish failed: \(error.localizedDescription)", jobId: uploadId)
                                completion(.failure(error))
                            }
                        }
                    },
                    progress: { bytes, total in
                        onProgress(bytes, total)
                    }
                )
            } catch {
                TelemetryManager.shared.trackUploadFailure(error: "Background start failed: \(error.localizedDescription)")
                completion(.failure(error))
            }
        }
    }

    // MARK: - Upload

    func uploadFile(file: URL, onProgress: @escaping (Int64, Int64) -> Void) async throws -> UploadFinishResponse {
        let totalSize = try file.resourceValues(forKeys: [.fileSizeKey]).fileSize ?? 0
        let existing = try await fetchTransferState(filePath: file.path)
        let startTime = Date()

        let uploadId: String
        let chunkSize: Int64

        if let existing = existing, let existingUploadId = existing.uploadId {
            uploadId = existingUploadId
            chunkSize = self.chunkSize
            print("Resuming upload for \(file.lastPathComponent) from \(existing.bytesTransferred) bytes")
        } else {
            do {
                let start = try await api.startResumableUpload(
                    filename: file.lastPathComponent,
                    filepath: file.path,
                    totalSize: Int64(totalSize)
                )
                uploadId = start.uploadId
                chunkSize = start.chunkSize
                let state = TransferState(context: context)
                state.id = UUID()
                state.uploadId = uploadId
                state.filePath = file.path
                state.totalBytes = Int64(totalSize)
                state.status = TransferStatus.pending.rawValue
                state.transferType = TransferType.upload.rawValue
                state.createdAt = Date()
                state.updatedAt = Date()
                PersistenceController.shared.save()
                TelemetryManager.shared.trackUploadStart(jobId: uploadId)
            } catch {
                let duration = Int64(Date().timeIntervalSince(startTime) * 1000)
                TelemetryManager.shared.trackUploadFailure(error: "Start failed: \(error.localizedDescription)", durationMs: duration)
                throw error
            }
        }

        var offset = existing?.bytesTransferred ?? 0
        try await updateStatus(uploadId: uploadId, status: .active, bytesTransferred: offset)

        let handle = try FileHandle(forReadingFrom: file)
        defer { try? handle.close() }

        do {
            while offset < Int64(totalSize) {
                let end = min(offset + chunkSize - 1, Int64(totalSize) - 1)
                try handle.seek(toOffset: UInt64(offset))
                let chunk = handle.readData(ofLength: Int(end - offset + 1))

                let result = try await retryWithBackoff { attempt in
                    try await api.uploadChunk(
                        uploadId: uploadId,
                        start: offset,
                        end: end,
                        total: Int64(totalSize),
                        data: chunk
                    )
                }

                offset = end + 1
                try await updateStatus(uploadId: uploadId, status: .active, bytesTransferred: offset)
                onProgress(offset, Int64(totalSize))
            }

            let finish = try await retryWithBackoff { _ in
                try await api.finishResumableUpload(uploadId: uploadId)
            }

            if let state = try await fetchTransferState(uploadId: uploadId) {
                context.delete(state)
                PersistenceController.shared.save()
            }

            let duration = Int64(Date().timeIntervalSince(startTime) * 1000)
            TelemetryManager.shared.trackUploadSuccess(jobId: uploadId, durationMs: duration, bytesTransferred: Int64(totalSize))
            NotificationManager.shared.showTransferComplete(
                title: "Upload Complete",
                body: "\(file.lastPathComponent) queued for processing"
            )

            return finish
        } catch {
            let duration = Int64(Date().timeIntervalSince(startTime) * 1000)
            TelemetryManager.shared.trackUploadFailure(error: "Upload failed: \(error.localizedDescription)", jobId: uploadId, durationMs: duration)
            throw error
        }
    }

    // MARK: - Download

    func downloadFile(jobId: String, outputFile: URL, totalSize: Int64, onProgress: @escaping (Int64, Int64) -> Void) async throws {
        let tempFile = outputFile.deletingLastPathComponent().appendingPathComponent(outputFile.lastPathComponent + ".part")
        let existing = try await fetchTransferState(jobId: jobId)
        var offset = existing?.bytesTransferred ?? 0
        let startTime = Date()

        if FileManager.default.fileExists(atPath: tempFile.path) && offset == 0 {
            let attrs = try FileManager.default.attributesOfItem(atPath: tempFile.path)
            offset = attrs[.size] as? Int64 ?? 0
        }

        let state = TransferState(context: context)
        state.id = UUID()
        state.jobId = jobId
        state.filePath = outputFile.path
        state.localTempPath = tempFile.path
        state.totalBytes = totalSize
        state.bytesTransferred = offset
        state.status = TransferStatus.active.rawValue
        state.transferType = TransferType.download.rawValue
        state.createdAt = Date()
        state.updatedAt = Date()
        PersistenceController.shared.save()
        TelemetryManager.shared.trackDownloadStart(jobId: jobId)

        let handle = try FileHandle(forWritingTo: tempFile)
        defer { try? handle.close() }

        do {
            while offset < totalSize {
                let end = min(offset + chunkSize - 1, totalSize - 1)
                let range = "bytes=\(offset)-\(end)"

                let data = try await retryWithBackoff { _ in
                    try await api.downloadOutput(jobId: jobId, range: range)
                }

                try handle.seek(toOffset: UInt64(offset))
                try handle.write(contentsOf: data)
                offset += Int64(data.count)
                try await updateStatus(jobId: jobId, status: .active, bytesTransferred: offset)
                onProgress(offset, totalSize)
            }

            try FileManager.default.moveItem(at: tempFile, to: outputFile)
            if let state = try await fetchTransferState(jobId: jobId) {
                context.delete(state)
                PersistenceController.shared.save()
            }

            let duration = Int64(Date().timeIntervalSince(startTime) * 1000)
            TelemetryManager.shared.trackDownloadSuccess(jobId: jobId, durationMs: duration, bytesTransferred: totalSize)
            NotificationManager.shared.showTransferComplete(
                title: "Download Complete",
                body: "Saved to \(outputFile.lastPathComponent)"
            )
        } catch {
            let duration = Int64(Date().timeIntervalSince(startTime) * 1000)
            TelemetryManager.shared.trackDownloadFailure(error: "Download failed: \(error.localizedDescription)", jobId: jobId, durationMs: duration)
            throw error
        }
    }

    // MARK: - Retry

    private func retryWithBackoff<T>(operation: (Int) async throws -> T) async throws -> T {
        var lastError: Error?
        for attempt in 0..<maxRetries {
            do {
                return try await operation(attempt)
            } catch {
                lastError = error
                let delay = calculateDelay(attempt: attempt)
                print("Retry \(attempt)/\(maxRetries) after \(delay)ms: \(error)")
                try await Task.sleep(nanoseconds: UInt64(delay * 1_000_000))
            }
        }
        NotificationManager.shared.showTransferError(
            title: "Transfer Failed",
            body: "Max retries exceeded: \(lastError?.localizedDescription ?? "Unknown error")"
        )
        throw EnkoduError.maxRetriesExceeded
    }

    private func calculateDelay(attempt: Int) -> Double {
        let base = baseDelayMs * pow(backoffMultiplier, Double(attempt))
        let jitter = base * 0.3 * Double.random(in: 0...1)
        return min(base + jitter, maxDelayMs)
    }

    // MARK: - State management

    private func fetchTransferState(uploadId: String? = nil, jobId: String? = nil, filePath: String? = nil) async throws -> TransferState? {
        let request = TransferState.fetchRequest()
        if let uploadId = uploadId {
            request.predicate = NSPredicate(format: "uploadId == %@", uploadId)
        } else if let jobId = jobId {
            request.predicate = NSPredicate(format: "jobId == %@", jobId)
        } else if let filePath = filePath {
            request.predicate = NSPredicate(format: "filePath == %@", filePath)
        }
        request.fetchLimit = 1
        return try context.fetch(request).first
    }

    private func updateStatus(uploadId: String? = nil, jobId: String? = nil, status: TransferStatus, bytesTransferred: Int64) async throws {
        let state = try await fetchTransferState(uploadId: uploadId, jobId: jobId)
        state?.status = status.rawValue
        state?.bytesTransferred = bytesTransferred
        state?.updatedAt = Date()
        PersistenceController.shared.save()
    }

    // MARK: - Pause / Resume

    func pauseAll() {
        let request = TransferState.fetchRequest()
        request.predicate = NSPredicate(format: "status == %@", TransferStatus.active.rawValue)
        do {
            let active = try context.fetch(request)
            for state in active {
                state.status = TransferStatus.paused.rawValue
                state.updatedAt = Date()
            }
            PersistenceController.shared.save()
        } catch {
            print("Failed to pause transfers: \(error)")
        }
    }

    func resumeAll() {
        // Resume logic would be triggered by the UI
        // This just marks paused transfers as pending for resumption
        let request = TransferState.fetchRequest()
        request.predicate = NSPredicate(format: "status == %@", TransferStatus.paused.rawValue)
        do {
            let paused = try context.fetch(request)
            for state in paused {
                state.status = TransferStatus.pending.rawValue
                state.updatedAt = Date()
            }
            PersistenceController.shared.save()
        } catch {
            print("Failed to resume transfers: \(error)")
        }
    }
}

File: mobile/android/app/src/main/java/com/enkodu/companion/transfer/TransferManager.kt
package com.enkodu.companion.transfer

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.util.Log
import com.enkodu.companion.api.ChunkResponse
import com.enkodu.companion.api.EnkoduApi
import com.enkodu.companion.api.ResumableStartRequest
import com.enkodu.companion.api.ResumableStartResponse
import com.enkodu.companion.api.UploadFinishResponse
import com.enkodu.companion.data.TransferDao
import com.enkodu.companion.data.TransferState
import com.enkodu.companion.data.TransferStatus
import com.enkodu.companion.data.TransferType
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import retrofit2.HttpException
import java.io.File
import java.io.RandomAccessFile
import java.io.IOException
import kotlin.math.min
import kotlin.math.pow
import kotlin.random.Random

class TransferManager(
    private val context: Context,
    private val api: EnkoduApi,
    private val dao: TransferDao,
    private val telemetry: com.enkodu.companion.TelemetryManager? = null
) {
    companion object {
        private const val TAG = "TransferManager"
        private const val CHUNK_SIZE = 8 * 1024 * 1024L // 8 MiB
        private const val MAX_RETRIES = 10
        private const val BASE_DELAY_MS = 500L
        private const val MAX_DELAY_MS = 30000L
        private const val BACKOFF_MULTIPLIER = 1.5
    }

    sealed class Result<out T> {
        data class Success<T>(val data: T) : Result<T>()
        data class Failure(val error: String, val retryable: Boolean = true) : Result<Nothing>()
    }

    // ── Upload ────────────────────────────────────────────────────────────────

    suspend fun uploadFile(
        file: File,
        onProgress: (bytesTransferred: Long, totalBytes: Long) -> Unit
    ): Result<UploadFinishResponse> = withContext(Dispatchers.IO) {
        val totalSize = file.length()
        val existing = dao.getByFilePath(file.absolutePath)
        val startTime = System.currentTimeMillis()

        val uploadId: String
        val chunkSize: Long

        if (existing != null && existing.uploadId != null) {
            uploadId = existing.uploadId
            chunkSize = CHUNK_SIZE
            Log.i(TAG, "Resuming upload for ${file.name} from ${existing.bytesTransferred} bytes")
        } else {
            val startResp = startUpload(file, totalSize)
            when (startResp) {
                is Result.Success -> {
                    uploadId = startResp.data.uploadId
                    chunkSize = startResp.data.chunkSize
                    dao.insert(
                        TransferState(
                            uploadId = uploadId,
                            filePath = file.absolutePath,
                            totalBytes = totalSize,
                            transferType = TransferType.UPLOAD.name,
                            status = TransferStatus.PENDING.name
                        )
                    )
                    telemetry?.trackUploadStart(uploadId)
                }
                is Result.Failure -> {
                    telemetry?.trackUploadFailure("Start failed: ${startResp.error}", durationMs = System.currentTimeMillis() - startTime)
                    return@withContext Result.Failure(startResp.error, startResp.retryable)
                }
            }
        }

        val transferred = existing?.bytesTransferred ?: 0
        dao.updateProgress(uploadId, transferred, TransferStatus.ACTIVE.name)

        RandomAccessFile(file, "r").use { raf ->
            var offset = transferred
            while (offset < totalSize) {
                val end = min(offset + chunkSize - 1, totalSize - 1)
                val chunkSizeActual = (end - offset + 1).toInt()
                val chunk = ByteArray(chunkSizeActual)
                raf.seek(offset)
                raf.readFully(chunk)

                val chunkResult = retryWithBackoff { attempt ->
                    val contentRange = "bytes $offset-$end/$totalSize"
                    val body = chunk.toRequestBody("application/octet-stream".toMediaType())
                    val resp = api.uploadChunk(uploadId, contentRange, body)
                    if (resp.isSuccessful) {
                        resp.body()?.let { Result.Success(it) }
                            ?: Result.Failure("Empty chunk response", true)
                    } else {
                        Result.Failure("Chunk failed: ${resp.code()}", isRetryable(resp.code()))
                    }
                }

                when (chunkResult) {
                    is Result.Success -> {
                        offset = end + 1
                        dao.updateProgress(uploadId, offset, TransferStatus.ACTIVE.name)
                        onProgress(offset, totalSize)
                    }
                    is Result.Failure -> {
                        val duration = System.currentTimeMillis() - startTime
                        when {
                            chunkResult.error.contains("401") -> {
                                dao.updateProgress(uploadId, offset, TransferStatus.PAUSED.name)
                                telemetry?.trackUploadFailure("auth_401 at $offset: ${chunkResult.error}", uploadId, duration)
                                return@withContext Result.Failure("auth_401: token missing or rejected", retryable = false)
                            }
                            chunkResult.error.contains("403") -> {
                                dao.updateProgress(uploadId, offset, TransferStatus.FAILED.name)
                                telemetry?.trackUploadFailure("auth_403 at $offset: ${chunkResult.error}", uploadId, duration)
                                return@withContext Result.Failure("auth_403: permission denied", retryable = false)
                            }
                            else -> {
                                dao.updateProgress(uploadId, offset, TransferStatus.PAUSED.name)
                                telemetry?.trackUploadFailure("Paused at $offset: ${chunkResult.error}", uploadId, duration)
                                return@withContext Result.Failure(
                                    "Upload paused at $offset: ${chunkResult.error}",
                                    chunkResult.retryable
                                )
                            }
                        }
                    }
                }
            }
        }

        // Finish upload
        val finishResult = retryWithBackoff { _ ->
            val resp = api.finishResumableUpload(uploadId)
            if (resp.isSuccessful) {
                resp.body()?.let { Result.Success(it) }
                    ?: Result.Failure("Empty finish response", true)
            } else {
                Result.Failure("Finish failed: ${resp.code()}", isRetryable(resp.code()))
            }
        }

        val duration = System.currentTimeMillis() - startTime
        when (finishResult) {
            is Result.Success -> {
                dao.deleteByUploadId(uploadId)
                telemetry?.trackUploadSuccess(uploadId, duration, totalSize)
                Result.Success(finishResult.data)
            }
            is Result.Failure -> {
                when {
                    finishResult.error.contains("401") -> {
                        dao.updateProgress(uploadId, totalSize, TransferStatus.PAUSED.name)
                        telemetry?.trackUploadFailure("Finish auth_401: ${finishResult.error}", uploadId, duration)
                        Result.Failure("auth_401: token missing or rejected", retryable = false)
                    }
                    finishResult.error.contains("403") -> {
                        dao.updateProgress(uploadId, totalSize, TransferStatus.FAILED.name)
                        telemetry?.trackUploadFailure("Finish auth_403: ${finishResult.error}", uploadId, duration)
                        Result.Failure("auth_403: permission denied", retryable = false)
                    }
                    else -> {
                        dao.updateProgress(uploadId, totalSize, TransferStatus.FAILED.name)
                        telemetry?.trackUploadFailure("Finish failed: ${finishResult.error}", uploadId, duration)
                        Result.Failure(finishResult.error, finishResult.retryable)
                    }
                }
            }
        }
    }

    private suspend fun startUpload(file: File, totalSize: Long): Result<ResumableStartResponse> {
        return retryWithBackoff { _ ->
            val resp = api.startResumableUpload(
                ResumableStartRequest(
                    filename = file.name,
                    filepath = file.absolutePath,
                    totalSize = totalSize
                )
            )
            if (resp.isSuccessful) {
                resp.body()?.let { Result.Success(it) }
                    ?: Result.Failure("Empty start response", true)
            } else {
                Result.Failure("Start failed: ${resp.code()}", isRetryable(resp.code()))
            }
        }
    }

    // ── Download ──────────────────────────────────────────────────────────────

    suspend fun downloadFile(
        jobId: String,
        outputFile: File,
        totalSize: Long,
        onProgress: (bytesTransferred: Long, totalBytes: Long) -> Unit
    ): Result<File> = withContext(Dispatchers.IO) {
        val tempFile = File(outputFile.parent, "${outputFile.name}.part")
        val existing = dao.getByJobId(jobId)
        var offset = existing?.bytesTransferred ?: 0
        val startTime = System.currentTimeMillis()

        if (tempFile.exists() && offset == 0L) {
            offset = tempFile.length()
        }

        Log.i(TAG, "Downloading $jobId to ${outputFile.name} from offset $offset")

        dao.insert(
            TransferState(
                jobId = jobId,
                filePath = outputFile.absolutePath,
                localTempPath = tempFile.absolutePath,
                totalBytes = totalSize,
                bytesTransferred = offset,
                transferType = TransferType.DOWNLOAD.name,
                status = TransferStatus.ACTIVE.name
            )
        )
        telemetry?.trackDownloadStart(jobId)

        RandomAccessFile(tempFile, "rw").use { raf ->
            raf.setLength(totalSize)
            while (offset < totalSize) {
                val end = min(offset + CHUNK_SIZE - 1, totalSize - 1)
                val range = "bytes=$offset-$end"

                val chunkResult = retryWithBackoff { _ ->
                    val resp = api.downloadOutput(jobId, range)
                    if (resp.isSuccessful) {
                        val body = resp.body()
                        if (body != null) {
                            val bytes = body.bytes()
                            Result.Success(bytes)
                        } else {
                            Result.Failure("Empty download body", true)
                        }
                    } else {
                        Result.Failure("Download failed: ${resp.code()}", isRetryable(resp.code()))
                    }
                }

                when (chunkResult) {
                    is Result.Success -> {
                        raf.seek(offset)
                        raf.write(chunkResult.data)
                        offset += chunkResult.data.size
                        dao.updateProgressByJobId(jobId, offset, TransferStatus.ACTIVE.name)
                        onProgress(offset, totalSize)
                    }
                    is Result.Failure -> {
                        val duration = System.currentTimeMillis() - startTime
                        when {
                            chunkResult.error.contains("401") -> {
                                dao.updateProgressByJobId(jobId, offset, TransferStatus.PAUSED.name)
                                telemetry?.trackDownloadFailure("auth_401 at $offset: ${chunkResult.error}", jobId, duration)
                                return@withContext Result.Failure("auth_401: token missing or rejected", retryable = false)
                            }
                            chunkResult.error.contains("403") -> {
                                dao.updateProgressByJobId(jobId, offset, TransferStatus.FAILED.name)
                                telemetry?.trackDownloadFailure("auth_403 at $offset: ${chunkResult.error}", jobId, duration)
                                return@withContext Result.Failure("auth_403: permission denied", retryable = false)
                            }
                            else -> {
                                dao.updateProgressByJobId(jobId, offset, TransferStatus.PAUSED.name)
                                telemetry?.trackDownloadFailure("Paused at $offset: ${chunkResult.error}", jobId, duration)
                                return@withContext Result.Failure(
                                    "Download paused at $offset: ${chunkResult.error}",
                                    chunkResult.retryable
                                )
                            }
                        }
                    }
                }
            }
        }

        // Guard: only save if job is verified
        val jobResp = try { api.getJob(jobId) } catch (e: Exception) { null }
        val job = jobResp?.body()
        if (job == null || job.status != "done" || job.verifyStatus != "pass") {
            tempFile.delete()
            val reason = when {
                job == null -> "job_status_unknown"
                job.status != "done" -> "job_not_done"
                else -> "verify_status_not_pass: ${job.verifyStatus}"
            }
            dao.updateProgressByJobId(jobId, 0L, TransferStatus.FAILED.name)
            telemetry?.trackDownloadFailure("Verification gate: $reason", jobId, System.currentTimeMillis() - startTime)
            return@withContext Result.Failure("verify_gate: $reason", retryable = false)
        }

        // Checksum verification
        val checksumResp = try { api.getChecksum(jobId) } catch (e: Exception) { null }
        val expectedSha = checksumResp?.body()?.outputSha256
        if (expectedSha != null) {
            val actualSha = computeSha256(tempFile)
            if (!actualSha.equals(expectedSha, ignoreCase = true)) {
                tempFile.delete()
                dao.updateProgressByJobId(jobId, 0L, TransferStatus.FAILED.name)
                telemetry?.trackDownloadFailure("checksum_mismatch", jobId, System.currentTimeMillis() - startTime)
                return@withContext Result.Failure("checksum_mismatch", retryable = false)
            }
        }

        tempFile.renameTo(outputFile)
        dao.deleteByJobId(jobId)
        val duration = System.currentTimeMillis() - startTime
        telemetry?.trackDownloadSuccess(jobId, duration, totalSize)
        Result.Success(outputFile)
    }

    // ── Retry logic ───────────────────────────────────────────────────────────

    private suspend fun <T> retryWithBackoff(
        operation: suspend (attempt: Int) -> Result<T>
    ): Result<T> {
        var lastError = ""
        for (attempt in 0..MAX_RETRIES) {
            when (val result = operation(attempt)) {
                is Result.Success -> return result
                is Result.Failure -> {
                    lastError = result.error
                    if (!result.retryable) return result
                    if (attempt < MAX_RETRIES) {
                        val delay = calculateDelay(attempt)
                        Log.d(TAG, "Retry $attempt/$MAX_RETRIES after ${delay}ms: $lastError")
                        delay(delay)
                    }
                }
            }
        }
        return Result.Failure("Max retries exceeded: $lastError", false)
    }

    private fun calculateDelay(attempt: Int): Long {
        val base = BASE_DELAY_MS * BACKOFF_MULTIPLIER.pow(attempt.toDouble())
        val jitter = base * 0.3 * Random.nextDouble()
        return min((base + jitter).toLong(), MAX_DELAY_MS)
    }

    private fun isRetryable(statusCode: Int): Boolean {
        return statusCode in listOf(408, 429, 500, 502, 503, 504)
    }

    private fun isAuthError(statusCode: Int): Boolean = statusCode == 401 || statusCode == 403

    // ── Constraints ─────────────────────────────────────────────────────────────

    fun isWifiConnected(): Boolean {
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = cm.activeNetwork ?: return false
        val capabilities = cm.getNetworkCapabilities(network) ?: return false
        return capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
    }

    fun isCellularConnected(): Boolean {
        val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val network = cm.activeNetwork ?: return false
        val capabilities = cm.getNetworkCapabilities(network) ?: return false
        return capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)
    }

    private fun computeSha256(file: File): String {
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        file.inputStream().use { input ->
            val buffer = ByteArray(8192)
            var read: Int
            while (input.read(buffer).also { read = it } != -1) {
                digest.update(buffer, 0, read)
            }
        }
        return digest.digest().joinToString("") { "%02x".format(it) }
    }
}
